qpmodel/optimizer.cs

```diff
@@ -50,8 +51,61 @@ public class Property { }
     public class PhysicProperty : Property
     {
         // ordering: the ordered expression and whether is descending
-        KeyValuePair<Expr, bool> ordering_;
-        KeyValuePair<Expr, DistributionType> distribution_;
+        List<KeyValuePair<Expr, bool>> ordering_;
+        public PhysicProperty(List<Expr> order, List<bool> desc = null)
+        {
+            if (desc is null)
+                desc = new List<bool>(Enumerable.Repeat(false, order.Count).ToArray());
+            Debug.Assert(order.Count == desc.Count);
+            ordering_ = new List<KeyValuePair<Expr, bool>>();
+            for(int i=0; i<order.Count; i++)
+                ordering_.Add(new KeyValuePair<Expr, bool> (order[i], desc[i]));
+        }
+        public bool IsPropertySupplied(PhysicNode node)
+        {
+            if (this.Equals(node.SuppiedProperty()))
+                return true;
+            return false;
+        }
+
+        public CGroupMember EnforceProperty(CMemoGroup group)
+        {
+            var physic = this.OrderEnforcement(group.minMember_.physic_);
+            return new CGroupMember(physic, group);
+        }
+        internal PhysicNode OrderEnforcement(PhysicNode node)
+        {
+            List<Expr> order = new List<Expr>();
+            List<bool> desc = new List<bool>();
+            foreach (var pair in ordering_)
+            {
+                order.Add(pair.Key);
+                desc.Add(pair.Value);
+            }
+            var logicnode = new LogicOrder(node.logic_, order, desc);
+            return new PhysicOrder(logicnode, node);
+        }
+
+        public bool Equals(PhysicProperty other)
+        {
+            if (other is null) return false;
+            if (ordering_.Count != other.ordering_.Count) return false;
+            for (int i = 0; i < ordering_.Count; i++)
+            {
+                if (!ordering_[i].Key.Equals(other.ordering_[i].Key)) return false;
+                if (ordering_[i].Value != other.ordering_[i].Value) return false;
+            }
+            return true;
+        }
+        public override bool Equals(object obj)
+        {
+            if (obj.GetType() != this.GetType()) return false;
```

Comment:
if (!(obj is PhysicProperty)) return false?

Reply:
I wanted to declare that if if is compared to a object that is not PhysicProperty then they cannot be equal.